openzim/cms
# Add query filtering parameters for `/titles`

Closes #41 

This is a work-in-progress.

@rgaudin 

I have failing BUT valid tests in this PR.

We will use/need `exclude_unset` refer this https://pydantic-docs.helpmanual.io/usage/exporting_models/
 whether fields which were not explicitly set when creating the model should be excluded from the returned dictionary; default False


I am still confused how they are using it here in FastApi https://fastapi.tiangolo.com/tutorial/body-updates/#using-pydantics-exclude_unset-parameter

here's ORMar's same docs for unsetting https://collerek.github.io/ormar/models/methods/#exclude_unset

---

following is my notes on this issue

-  [x] https://github.com/openzim/cms/issues/41
	- filters
		- [ ]  `?metadata-name=wikipedia_fr_test` to filter based on a meta (`Name` here). Value can include the `*` character (ex: `wikipedia_*_maxi`, `*_nopic`, etc). `Illustration_*` are not supported obviously.
			- [ ] exact match, `?metadata-name=wikipedia_fr_test`
			- [ ] wildcard match, `?metadata-name=wikipedia_*_all`
				- look into **SELECT LIKE**
			- [ ] ignore `Illustration_*`
		- [ ] `?tag=wikipedia,kids`, `?tag=wikipedia|wikibooks`
- [ ] payload-control
	- [ ] `?with_metadata=title,scraper,illustration_48x48`: includes the requested metadata in response. `all` is a shortcut for including all metadata (including illustration). `all_text` is same but only text-ones. `illustration` is an alias for `illustration_48x48`.
	- [ ] **`?with_languages`: include languages into response**
		- the `?with_languages` just by itself is not supported, it needs some value
			- ```diff
				-?with_languages
				+?with_languages=true
		- test
		- add `False` as the default
		- use the `bool` type
		- ❌ the `fields()` is not having any effect
		- test cases
			- [x] titles with single language
			- [x] titles with multiple languages
		- ⭕ Should all the `language` response follow the above style, instead of `["eng", "fra"]`
	- [ ] `?with_tags`: include tags (all) into response.
		- test cases
			- [x] combine `with_languages` and `with_tags`

**Meeting with Renaud** at 17:30
- Q. Will we ever have Titles w/o languages?
	- No
- Q. Should we be sending language data of titles by default. Because `fields("ident")` is not able to remove languages when `select_related()` is used
	- unless `with_languages` is used, don't send "language" key in the response
- Q. Will there be test case where `?lang=fra&with_languages=false`?
	- Yes

**Notes**
- We could just not use pydantic `response_model` for parsing, and then we can customise the response. Note of the pagination working with this new custom response. Refer `/titles/{ident}` on how to do this.
- Code is becoming spaghetti when I try to get multiple query parameters to work together, a.k.a. complex request
- Maybe if we use the difference between `None` vs `[]`, use can still use Pydantic schema?
